File: samples/cpp/SampleApp/src/TextToSpeech/TextToSpeechHandler.cpp
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include "SampleApp/TextToSpeech/TextToSpeechHandler.h"
// C++ Standard Library
#include <sstream>
#include <unistd.h>
#include <utility>
#include <fstream>

// Guidelines Support Library
#define GSL_THROW_ON_CONTRACT_VIOLATION
#include <gsl/contracts.h>

namespace sampleApp {
namespace textToSpeech {

static constexpr size_t READ_BUFFER_SIZE = 4096;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//  TextToSpeechHandler
//
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
static const std::string ID = "TEXT_TO_SPEECH";
static const std::string TEXT_TO_SPEECH_PROVIDER = "text-to-speech-provider";
using json = nlohmann::json;
TextToSpeechHandler::TextToSpeechHandler(
    std::weak_ptr<Activity> activity,
    std::weak_ptr<logger::LoggerHandler> loggerHandler,
    std::weak_ptr<audio::AudioOutputProviderHandler> audioOutputProviderHandler) :
        m_activity{std::move(activity)},
        m_loggerHandler{std::move(loggerHandler)},
        m_audioOutputHandler{std::move(audioOutputProviderHandler)},
        m_idCounter(0) {
    setupUI();
}

std::weak_ptr<logger::LoggerHandler> TextToSpeechHandler::getLoggerHandler() {
    return m_loggerHandler;
}

std::weak_ptr<Activity> TextToSpeechHandler::getActivity() {
    return m_activity;
}

// aace::textToSpeech:TextToSpeech interface
void TextToSpeechHandler::prepareSpeechCompleted(
    const std::string& speechId,
    std::shared_ptr<aace::audio::AudioStream> preparedAudio,
    const std::string& metadata) {
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }
    activity->runOnUIThread([=]() {
        if (auto console = m_console.lock()) {
            console->printRuler();
            console->printLine("Prepare speech Completed");
            console->printLine("Speech ID      : ", speechId);
            console->printLine("Metadata       : ", metadata);
            console->printRuler();
        }
    });
    // For QA testing only. The OEM application should follow the UX guidelines in
    // order to play the audio
    try {
        auto path = "" + speechId + ".mp3";
        log(logger::LoggerHandler::Level::VERBOSE, "path" + path);

        // create the output file
        auto output = std::make_shared<std::ofstream>(path, std::ios::binary);
        // copy the stream to the file
        char buffer[READ_BUFFER_SIZE];
        ssize_t size = 0;

        while (!preparedAudio->isClosed()) {
            ssize_t bytesRead = preparedAudio->read(buffer, READ_BUFFER_SIZE);

            // throw an error if the read failed
            if (bytesRead < 0) {
                log(logger::LoggerHandler::Level::VERBOSE, "bytes read is zero");
                return;
            }

            // write the data to the output file
            output->write(buffer, bytesRead);

            size += bytesRead;
        }

        output->close();
        log(logger::LoggerHandler::Level::VERBOSE, "complete size" + size);
    } catch (std::exception& ex) {
        log(logger::LoggerHandler::Level::ERROR, ex.what());
    }
}

void TextToSpeechHandler::prepareSpeechFailed(const std::string& speechId, const std::string& reason) {
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }
    activity->runOnUIThread([=]() {
        if (auto console = m_console.lock()) {
            console->printRuler();
            console->printLine("Prepare speech Failed");
            console->printLine("Speech ID      : ", speechId);
            console->printLine("Reason       : ", reason);
            console->printRuler();
        }
    });
}

void TextToSpeechHandler::capabilitiesReceived(const std::string& requestId, const std::string& capabilities) {
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }
    activity->runOnUIThread([=]() {
        if (auto console = m_console.lock()) {
            console->printRuler();
            console->printLine("Capabilities Received");
            console->printLine("Capabilities      : ", capabilities);
            console->printLine("Request ID      : ", requestId);
            console->printRuler();
        }
    });
}

// private

void TextToSpeechHandler::log(logger::LoggerHandler::Level level, const std::string& message) {
    auto loggerHandler = m_loggerHandler.lock();
    if (!loggerHandler) {
        return;
    }
    loggerHandler->log(level, "TextToSpeechHandler", message);
}

void TextToSpeechHandler::setupUI() {
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }

    m_console = activity->findViewById("id:console");

    // prepareSpeech
    activity->registerObserver(Event::onPrepareSpeech, [=](const std::string& value) {
        auto speechId = generateId();
        log(logger::LoggerHandler::Level::VERBOSE, "onPrepareSpeech text: " + value + " speech ID: " + speechId);
        return prepareSpeech(speechId, value, TEXT_TO_SPEECH_PROVIDER, "");
    });

    // getCapabilities
    activity->registerObserver(Event::onGetCapabilities, [=](const std::string& value) {
        auto requestId = generateId();
        log(logger::LoggerHandler::Level::VERBOSE, "onGetCapabilities request ID: " + requestId);
        return getCapabilities(requestId, TEXT_TO_SPEECH_PROVIDER);
    });
}

std::string TextToSpeechHandler::generateId() {
    return ID + "-" + std::to_string(++m_idCounter);
}

}  // namespace textToSpeech
}  // namespace sampleApp

File: samples/android/app/src/main/java/com/amazon/sampleapp/impl/TextToSpeech/TextToSpeechHandler.java
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazon.sampleapp.impl.TextToSpeech;

import android.app.Activity;
import android.content.Context;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Environment;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.Toast;

import com.amazon.aace.audio.AudioStream;
import com.amazon.aace.textToSpeech.TextToSpeech;
import com.amazon.sampleapp.FileUtils;
import com.amazon.sampleapp.R;
import com.amazon.sampleapp.impl.Logger.LoggerHandler;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * The handler provides a sample implementation of the TextToSpeech platform
 * interface.
 */
public class TextToSpeechHandler extends TextToSpeech {
    private final Activity mActivity;
    private final LoggerHandler mLogger;
    private final Context mContext;
    private static final String sTag = "TextToSpeech";
    private static final String ID = "TEXT_TO_SPEECH";
    private static final String TEXT_KEY = "text";
    private static final String SSML_KEY = "ssml";
    private static final String TEXT_TO_SPEECH_PROVIDER = "text-to-speech-provider";
    private long idCounter = 0;

    /**
     * Constructor for TextToSpeech handler.
     * @param activity The Android activity instance.
     * @param logger The logger instance.
     */

    public TextToSpeechHandler(Activity activity, LoggerHandler logger) {
        mActivity = activity;
        mLogger = logger;
        mLogger.postInfo(sTag, "Creating TextToSpeechHandler");
        mContext = mActivity.getApplicationContext();
        try {
            setupGUI();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void prepareSpeechCompleted(String speechId, AudioStream preparedAudio, String metadata) {
        mLogger.postInfo(sTag, "speechId : " + speechId + ", metadata : " + metadata);

        // This code is for testing purpose only. OEMs should follow the correct UX guidelines to
        // play the audio returned in this API.
        try (FileOutputStream os = mContext.openFileOutput("test-audio", Context.MODE_PRIVATE)) {
            byte[] buffer = new byte[4096];
            int size;
            while (!preparedAudio.isClosed()) {
                while ((size = preparedAudio.read(buffer)) > 0) os.write(buffer, 0, size);
            }
        } catch (Exception e) {
            mLogger.postError(sTag, e);
        }

        try {
            Uri uri = Uri.fromFile(mContext.getFileStreamPath("test-audio"));
            MediaPlayer mediaPlayer = new MediaPlayer();
            mediaPlayer.setDataSource(String.valueOf(uri));
            mediaPlayer.prepare();
            mediaPlayer.start();
        } catch (Exception e) {
            mLogger.postError(sTag, e.getMessage());
        }
    }

    @Override
    public void prepareSpeechFailed(String speechId, String reason) {
        mLogger.postInfo(sTag, "speechId : " + speechId + " reason : " + reason);
    }

    @Override
    public void capabilitiesReceived(String requestId, String capabilities) {
        mLogger.postInfo(sTag, "request Id" + requestId + "capabilities : " + capabilities);
    }

    public void setupGUI() {
        try {
            mLogger.postInfo(sTag, "Setting up UI...");

            final List<String> ttsTextOnlyArray = new ArrayList<String>(
                    Arrays.asList("", "Take Exit 39B", "Take the next left", "In 500ft, take the next exit",
                            "Wrong way, turn around", "You have reached your destination"));

            final List<String> ttsSSMLOnlyArray = new ArrayList<String>(Arrays.asList(
                    "", "<speak> Turn right on 3rd Lane <sub alias=\"Mercury\">Ln</sub> to merge </speak>"));

            // The TextToSpeechTestCases.json should be used for testing purposes only.
            // The OEMs are expected to call the prepareSpeech() API with the correct text/SSML string
            String sdCardPath = Environment.getExternalStorageDirectory().getAbsolutePath();
            String externalTTSFilePath = sdCardPath + "/TextToSpeechTestCases.json";
            if (new File(externalTTSFilePath).exists()) {
                JSONObject ttsTestCases = FileUtils.parseFileAsJSONObject(externalTTSFilePath);
                if (ttsTestCases.has(TEXT_KEY)) {
                    JSONArray textOnlyCases = ttsTestCases.getJSONArray(TEXT_KEY);
                    for (int i = 0; i < textOnlyCases.length(); i++) {
                        ttsTextOnlyArray.add(textOnlyCases.get(i).toString());
                    }
                }
                if (ttsTestCases.has(SSML_KEY)) {
                    JSONArray ssmlOnlyCases = ttsTestCases.getJSONArray(SSML_KEY);
                    for (int i = 0; i < ssmlOnlyCases.length(); i++) {
                        ttsSSMLOnlyArray.add(ssmlOnlyCases.get(i).toString());
                    }
                }
            }
            setUpTTSSpinner(ttsTextOnlyArray, TEXT_KEY);
            setUpTTSSpinner(ttsSSMLOnlyArray, SSML_KEY);
        } catch (Exception e) {
            mLogger.postError(sTag, e);
        }
    }

    public String generateId() {
        return String.valueOf(++idCounter);
    }

    public void setUpTTSSpinner(List<String> testCases, String type) {
        Spinner textToSpeechSpinner = null;
        if (type.equals("text")) {
            textToSpeechSpinner = (Spinner) mActivity.findViewById(R.id.textToSpeechTextOnlySpinner);
        } else if (type.equals("ssml")) {
            textToSpeechSpinner = (Spinner) mActivity.findViewById(R.id.textToSpeechSSMLOnlySpinner);
        }
        final String[] ttsStrings = new String[testCases.size()];
        ArrayAdapter<String> textToSpeechAdapter = new ArrayAdapter<String>(
                mActivity, android.R.layout.simple_spinner_item, testCases.toArray(ttsStrings));
        textToSpeechAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        textToSpeechSpinner.setAdapter(textToSpeechAdapter);
        textToSpeechSpinner.setSelection(0);

        textToSpeechSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                String ttsString = testCases.get(position);
                String speechId = ID + "-" + generateId();
                String options = "";
                Toast.makeText(mActivity, "Generating speech " + ttsString + "with Speech ID : " + speechId,
                             Toast.LENGTH_SHORT)
                        .show();
                prepareSpeech(speechId, ttsString, TEXT_TO_SPEECH_PROVIDER, options);
            }
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }
}